nlp_utilities.jl

```diff
@@ -0,0 +1,236 @@
+
+"""
+    all_params(model::Model)
+
+Get all the parameters in the model.
+"""
+all_params(model::Model) = filter(x -> is_parameter(x), all_variables(model))
+
+"""
+    create_evaluator(model::Model; x=all_variables(model))
+
+Create an evaluator for the model.
+"""
+function create_evaluator(model::Model; x=all_variables(model))
+    nlp, rows = create_nlp_model(model)
+    backend = MOI.Nonlinear.SparseReverseMode()
+    evaluator = MOI.Nonlinear.Evaluator(nlp, backend, index.(x))
+    MOI.initialize(evaluator, [:Hess, :Jac])
+    return evaluator, rows
+end
+
+"""
+    is_inequality(con::ConstraintRef)
+
+Check if the constraint is an inequality.
+"""
+function is_inequality(con::ConstraintRef)
+    set_type = typeof(MOI.get(owner_model(con), MOI.ConstraintSet(), con))
+    return set_type <: MOI.LessThan || set_type <: MOI.GreaterThan
+end
+
+"""
+    find_inequealities(cons::Vector{ConstraintRef})
+
+Find the indices of the inequality constraints.
+"""
+function find_inequealities(cons::Vector{ConstraintRef})
+    ineq_locations = zeros(length(cons))
+    for i in 1:length(cons)
+        if is_inequality(cons[i])
+            ineq_locations[i] = true
+        end
+    end
+    return findall(x -> x ==1, ineq_locations)
+end
+
+"""
+    get_slack_inequality(con::ConstraintRef)
+
+Get the reference to the canonical function that is equivalent to the slack variable of the inequality constraint.
+"""
+function get_slack_inequality(con::ConstraintRef)
+    slack = MOI.get(owner_model(con), CanonicalConstraintFunction(), con)
+    return slack
+end
+
+"""
+    compute_derivatives(evaluator::MOI.Nonlinear.Evaluator, cons::Vector{ConstraintRef}; primal_vars::Vector{VariableRef}=all_primal_vars(model), params::Vector{VariableRef}=all_params(model))
+
+Compute the derivatives of the solution with respect to the parameters.
+"""
+function compute_derivatives(evaluator::MOI.Nonlinear.Evaluator, cons::Vector{ConstraintRef}; 
+    primal_vars::Vector{VariableRef}, params::Vector{VariableRef}
+)
+    @assert all(x -> is_parameter(x), params) "All parameters must be parameters"
+
+    # Setting
+    num_vars = length(primal_vars)
+    num_parms = length(params)
+    num_cons = length(cons)
+    ineq_locations = find_inequealities(cons)
+    num_ineq = length(ineq_locations)
+    slack_vars = [get_slack_inequality(cons[i]) for i in ineq_locations]
+    all_vars = [primal_vars; params]
+
+    # Primal solution
+    X = diagm(value.([primal_vars; slack_vars]))
+
+    # Dual of the bounds
+    bound_duals = zeros(num_vars+num_ineq)
+    for i in 1:num_vars
+        if has_lower_bound(primal_vars[i])
+            bound_duals[i] = dual.(LowerBoundRef(primal_vars[i]))
+        end
+        if has_upper_bound(primal_vars[i])
+            bound_duals[i] -= dual.(UpperBoundRef(primal_vars[i]))
+        end
+    end
+    for (i, con) in enumerate(cons[ineq_locations])
+        bound_duals[num_vars+i] = dual.(con)
+    end
```

Comment:
Note that you're using JuMP's inequality multipliers for both the slack bound multipliers and the slack equality multipliers (in the Hessian). I believe this is correct, up to potentially a sign change. 

So as long as it's well-tested, I have no problem with using the slack reformulation, although I would personally find it simpler to just use the original problem's primal variables and inequality constraints.

Reply:
Since this is almost done, I think I will use the reformulation, test it, and then change to using the original problem (but now knowing my tests make sense).